frostsnap_core/src/coordinator.rs

```diff
@@ -3,485 +3,412 @@ use crate::{
-    }
-}
-
-#[derive(Clone, Debug)]
-pub enum CoordinatorState {
-    Registration,
-    KeyGen(KeyGenState),
-    FrostKey {
-        key: CoordinatorFrostKeyState,
-    },
-    Signing {
-        sign_state: SigningSessionState,
-        key: CoordinatorFrostKeyState,
-    },
-}
-
-#[derive(Clone, Debug, bincode::Encode, bincode::Decode)]
-pub struct SigningSessionState {
-    pub targets: BTreeSet<DeviceId>,
-    pub sessions: Vec<SignSessionProgress>,
-    pub request: SignRequest,
-}
-
-impl SigningSessionState {
-    pub fn received_from(&self) -> impl Iterator<Item = DeviceId> + '_ {
-        // all sessions make progress at the same time
-        self.sessions[0].received_from()
-    }
-}
-
-#[derive(Clone, Debug)]
-pub enum KeyGenState {
-    WaitingForResponses {
-        device_to_share_index: BTreeMap<DeviceId, Scalar<Public, NonZero>>,
-        responses: BTreeMap<DeviceId, Option<KeyGenResponse>>,
-        threshold: usize,
-    },
-    WaitingForAcks {
-        frost_key: FrostKey<Normal>,
-        device_to_share_index: BTreeMap<DeviceId, Scalar<Public, NonZero>>,
-        device_nonces: BTreeMap<DeviceId, DeviceNonces>,
-        acks: BTreeMap<DeviceId, bool>,
-        session_hash: SessionHash,
-    },
-}
-
-#[derive(Debug, Clone, bincode::Encode, bincode::Decode, serde::Serialize, serde::Deserialize)]
-pub struct DeviceNonces {
-    counter: usize,
-    nonces: VecDeque<Nonce>,
-}
-
-#[derive(Clone, Debug, bincode::Encode, bincode::Decode, serde::Serialize, serde::Deserialize)]
-pub struct CoordinatorFrostKeyState {
-    frost_key: FrostKey<Normal>,
-    device_to_share_index: BTreeMap<DeviceId, Scalar<Public, NonZero>>,
+#[derive(Debug, Clone, Default)]
+pub struct FrostCoordinator {
+    keys: BTreeMap<KeyId, CoordinatorFrostKey>,
+    key_order: Vec<KeyId>,
+    action_state: Option<CoordinatorState>,
     device_nonces: BTreeMap<DeviceId, DeviceNonces>,
 }
 
-impl CoordinatorFrostKeyState {
-    pub fn devices(&self) -> impl Iterator<Item = DeviceId> + '_ {
-        self.device_nonces.keys().cloned()
-    }
-
-    pub fn threshold(&self) -> usize {
-        self.frost_key.threshold()
-    }
-
-    pub fn frost_key(&self) -> &FrostKey<Normal> {
-        &self.frost_key
-    }
-
-    pub fn nonces_left(&self, device_id: DeviceId) -> Option<usize> {
-        let device_nonces = self.device_nonces.get(&device_id)?;
-        Some(device_nonces.nonces.len())
-    }
-}
-
 impl FrostCoordinator {
     pub fn new() -> Self {
         Self {
-            state: CoordinatorState::Registration,
+            keys: Default::default(),
+            key_order: Default::default(),
+            action_state: None,
+            device_nonces: Default::default(),
         }
     }
 
-    pub fn from_stored_key(key: CoordinatorFrostKeyState) -> Self {
-        Self {
-            state: CoordinatorState::FrostKey { key },
+    pub fn apply_change(&mut self, change: CoordinatorToStorageMessage) {
+        use CoordinatorToStorageMessage::*;
```

Comment:
Just for shorthanding the enum variants?

Reply:
Yes.